End-to-end tests for C++ wrapper of SPIFFE credentials

matthewstevenson88 opened the pull request:
This PR contains the end-to-end tests for the C++ wrapper of the SPIFFE credentials, which were merged in [PR 19778](https://github.com/grpc/grpc/pull/19778).

yihuazhang commented on include/grpcpp/security/tls_credentials_options.h:
```diff
@@ -56,7 +56,7 @@ class TlsKeyMaterialsConfig {
 
   /** Setter for key materials that will be called by the user. The setter
-   * transfers ownership of the arguments to the config. **/
+   *  takes ownership of the arguments to the config. **/
```
By looking at the implementation, it seems the statement does not hold for `set_pem_root_certs()`.

matthewstevenson88 replied:
Do you mean it doesn't hold for `add_pem_key_cert_pair`? If so, I agree: I forgot to add `std::move` there, and have done so in the next commit.

yihuazhang replied:
Sorry, Matt. I think I confused `add_pem_key_cert_pair` declared in both `TlsKeyMaterialsConfig` and `TlsCredentialReloadArg`, and mistakenly considered your comment applies to the `add_pem_key_cert_pair` defined in `TlsCredentialReloadArg`.

yihuazhang commented on test/cpp/util/spiffe_test_credentials.cc:
```diff
@@ -0,0 +1,84 @@
+
+class TestTlsServerAuthorizationCheckInterface
+    : public ::grpc_impl::experimental::TlsServerAuthorizationCheckInterface {
+  // Sync implementation.
```
Can we test async version of server authz check like we did in `h2_spiffe.cc`? By testing that, we will have a more confidence on our support to users' async callbacks.

yihuazhang replied:
Great! Thanks much for adding async test.

yihuazhang commented on test/cpp/util/spiffe_test_credentials.cc:
```diff
@@ -0,0 +1,84 @@
+};
+
+::grpc_impl::experimental::TlsCredentialsOptions*
+CreateTestTlsCredentialsOptions(bool is_client) {
```
Could you please add some comments within this function. It is a little hard to read the code.

matthewstevenson88 replied:
The next push will add extensive comments.

yihuazhang commented on test/cpp/util/spiffe_test_credentials.cc:
```diff
@@ -0,0 +1,84 @@
+  ::grpc_impl::experimental::TlsCredentialsOptions* options =
+      new ::grpc_impl::experimental::TlsCredentialsOptions(
+          is_client
+              ? GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE
```
I do not quite understand this part. First, this option should only be set at the server side (client will always verify server's certificates). So if we want to do mutual authentication, we only need to set server to `GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY`.

matthewstevenson88 replied:
In the next push, this flag is set to `GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY`. Note that, in the C++ wrapper, the form on the constructor for `TlsCredentialsOptions` forces the caller to set a cert request type.

yihuazhang commented on test/cpp/util/spiffe_test_credentials.h:
```diff
@@ -0,0 +1,59 @@
+ public:
+  std::shared_ptr<ChannelCredentials> GetChannelCredentials(
+      ChannelArguments* args) override {
+    args->SetSslTargetNameOverride("foo.test.google.fr");
```
Could you please add some comments here to explain why override is necessary?

matthewstevenson88 replied:
Added comments in the next push.

yihuazhang replied:
Thanks.

yihuazhang commented:
Also some failed tests (e.g., ASAN) are related to this PR.

matthewstevenson88 commented:
Thanks for the review @yihuazhang! The most recent commits address your comments, as well as do the following:

1. Add/expand unit tests and refactor the `SpiffeCredentialTypeProvider` API. Note that there is a hardcoded parameter in the e2e tests that force the test SPIFFE credentials to use async server authorization. Because of the testing framework (namely, the `CredentialProvider` interface), we can only have 1 SPIFFE test instance by default, so I opted for the async version.

2. Refactor the API's `TlsCredentials` and `TlsServerCredentials` to accept a shared pointer of `TlsCredentialsOptions` as input, rather than pass by constant reference. This brings the API's in line with the C core API's, and helps the caller better manage memory. (Note that changing the API is fine, since it lives in an experimental namespace.)

3. At the time of writing this comment, the ASAN test is not yet complete, but it passed when I ran it locally. So far there is a failing test (Bazel RBE Debug C/C++) that is failing because of the C core SPIFFE which we should maybe take a look at.

yihuazhang commented on src/core/lib/security/security_connector/ssl_utils.h:
```diff
@@ -86,6 +86,7 @@ grpc_security_status grpc_ssl_tsi_server_handshaker_factory_init(
 /* Exposed for testing only. */
 grpc_core::RefCountedPtr<grpc_auth_context> grpc_ssl_peer_to_auth_context(
     const tsi_peer* peer, const char* transport_security_type);
+
```
nit. Please remove the empty line.

yihuazhang commented on src/cpp/common/tls_credentials_options.cc:
```diff
@@ -274,6 +294,11 @@ TlsCredentialsOptions::TlsCredentialsOptions(
+ *  ownership of the c_credentials_options_ pointer (see e.g. the implementation
+ *  of the TlsCredentials API in secure_credentials.cc). For this reason, the
+ *  TlsCredentialsOptions destructor is not responsible for freeing
+ *  c_credentials_options_. **/
```
Since `c_credentials_options_` is created inside `TlsCredentialsOptions`constructor I think it also makes sense to destruct it in the destructor of `TlsCredentialsOptions`. Most of all, a caller does not know about `c_credentials_options_` right?

matthewstevenson88 replied:
That's true, but every current caller of `TlsCredentialsOptions` immediately takes ownership of `c_credentials_options_`. For that reason, if we try to delete `c_credentials_options_` in the destructor, then we will get an error.

yihuazhang replied:
I see. Can you please specify what are the callers of `TlsCredentialsOptions`? Do we have any callers other than users that create spiffe credentials?

matthewstevenson88 replied:
Outside of tests, the only callers are the `TlsCredentials` and `TlsServerCredentials` API's that are found in the files credentials.h, credentials_impl.h, server_credentials.h, and server_credentials_impl.h.

And yes, I can add this discussion and a list of the callers in the tls_credentials_options.h file where the destructor is exposed.

yihuazhang replied:
I see. Let's discuss more about it next Monday once you are at the office :)

yihuazhang commented on test/cpp/util/spiffe_test_credentials.h:
```diff
@@ -0,0 +1,128 @@
+ *
+ *  Further, the cert request type of the options instance is always set to
+ *  GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY for both the
+ *  client and the server. **/
```
I think `GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY` will only take effect at server side, not client side. Since client will always validate server's certificate.

matthewstevenson88 replied:
See my comment above.

yihuazhang commented on test/cpp/util/spiffe_test_credentials.cc:
```diff
@@ -0,0 +1,138 @@
+  }
+
+ private:
+  std::thread server_authz_check_thread_;
```
Can we define it as a function local variable instead of class member?

matthewstevenson88 replied:
Yes, that's probably better, to just define it locally in the `Schedule` function. Good call.

yihuazhang commented on test/cpp/util/spiffe_test_credentials.cc:
```diff
@@ -0,0 +1,138 @@
+ *      -otherwise, configured by TestSyncTlsServerAuthorizationCheck.
+ *  - otherwise, the server authorization check config is not populated.
+ *
+ *  Further, the cert request type of the options instance is always set to
```
I think GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY will only take effect at server side, not client side. Since client will always validate server's certificate.

matthewstevenson88 replied:
But I have to put some value for the client's cert request type, even if it doesn't do anything, right? For that reason, I just pass in `GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY` for both. If you think there's a better "default" option though, let me know! :)

yihuazhang replied:
For now you do not have to. But it may change in future (if we decide to support the feature that client also does not validate server's certificate).

matthewstevenson88 replied:
Sounds good. Thanks!

yihuazhang commented on test/cpp/util/spiffe_test_credentials.cc:
```diff
@@ -0,0 +1,138 @@
+   *  the parameters |is_client| and |is_async|. **/
+  std::shared_ptr<::grpc_impl::experimental::TlsServerAuthorizationCheckConfig>
+      test_server_authorization_check_config = nullptr;
+  if (is_client) {
```
Can we return early if `is_client` is `false` to reduce the nesting?

matthewstevenson88 replied:
Done.

yihuazhang commented:
Thanks much for adding async server authz test. I left some minor comments.

yihuazhang commented:
@matthewstevenson88 Is there any blocker for this PR? We have a request from OSS community wanting to know how to use TLSCredentials (e.g., credential reload and server authz).

matthewstevenson88 commented:
Sorry for the delay @yihuazhang! This should now be ready.

The only failing tests seem to be unrelated to this PR.

matthewstevenson88 commented:
FYI the one failing test seems to be related to the issue in PR #19898. (The log messages look slightly different, but it's the same task that is failing.)

yihuazhang commented on src/cpp/client/secure_credentials.cc:
```diff
@@ -280,9 +280,9 @@ std::shared_ptr<ChannelCredentials> LocalCredentials(
 // Builds TLS Credentials given TLS options.
 std::shared_ptr<ChannelCredentials> TlsCredentials(
-    const TlsCredentialsOptions& options) {
+    const std::shared_ptr<TlsCredentialsOptions>& options) {
```
Why do we need a `shared_ptr`? Credentials and options are usually 1:1 mapping

matthewstevenson88 replied:
Reverted back to constant reference in the next commit.

yihuazhang commented on src/cpp/common/tls_credentials_options.cc:
```diff
@@ -86,16 +83,39 @@ void TlsCredentialReloadArg::set_pem_root_certs(
+  c_arg_->key_materials_config->add_pem_key_cert_pair(c_pem_key_cert_pair);
+}
+
+void TlsCredentialReloadArg::set_key_materials(
```
nit. Please add some comments between lines.

matthewstevenson88 replied:
Added comments.

yihuazhang replied:
Thanks.

yihuazhang commented on include/grpcpp/security/tls_credentials_options.h:
```diff
@@ -103,7 +103,10 @@ class TlsCredentialReloadArg {
   void set_pem_root_certs(const grpc::string& pem_root_certs);
   void add_pem_key_cert_pair(
-      TlsKeyMaterialsConfig::PemKeyCertPair pem_key_cert_pair);
+      const TlsKeyMaterialsConfig::PemKeyCertPair& pem_key_cert_pair);
```
In the above comments - "They modify the fields of the underlying C arg...", it seems that they are are implementation details (i..e, C-core) that should not be exposed to the caller.

matthewstevenson88 replied:
I removed any reference to the "C arg" from the getters and setters (though I left one by the constructor that takes in a C arg as input).

yihuazhang commented on src/cpp/common/tls_credentials_options.cc:
```diff
@@ -86,16 +83,39 @@ void TlsCredentialReloadArg::set_pem_root_certs(
+  c_arg_->key_materials_config->add_pem_key_cert_pair(c_pem_key_cert_pair);
+}
+
+void TlsCredentialReloadArg::set_key_materials(
```
Please add some comments between lines for `set_key_materials` function.

matthewstevenson88 replied:
This was addressed in a previous comment (the one with "nit").

yihuazhang replied:
Thanks.

yihuazhang commented on src/cpp/server/secure_server_credentials.cc:
```diff
@@ -148,9 +148,9 @@ std::shared_ptr<ServerCredentials> LocalServerCredentials(
 
 std::shared_ptr<ServerCredentials> TlsServerCredentials(
-    const TlsCredentialsOptions& options) {
+    const std::shared_ptr<TlsCredentialsOptions>& options) {
```
Same question as above w.r.t `shared_ptr`.

matthewstevenson88 replied:
Changed to constant reference.

yihuazhang commented on test/cpp/client/credentials_test.cc:
```diff
@@ -53,8 +54,8 @@ static void tls_credential_reload_callback(
     GPR_ASSERT(arg != nullptr);
-    struct TlsKeyMaterialsConfig::PemKeyCertPair pair3 = {"private_key3",
-                                                          "cert_chain3"};
+    TlsKeyMaterialsConfig::PemKeyCertPair pair3 = {"private_key3",
```
s/pair3/pair

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/end2end/async_end2end_test.cc:
```diff
@@ -40,6 +40,7 @@
 #include "src/proto/grpc/testing/echo.grpc.pb.h"
 #include "test/core/util/port.h"
 #include "test/core/util/test_config.h"
+#include "test/cpp/util/spiffe_test_credentials.h"
```
maybe rename it to `tls_test_credentials`? as we have already renamed `spiffe` to `tls`.

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/client/credentials_test.cc:
```diff
@@ -660,14 +669,30 @@ TEST_F(CredentialsTest, LoadTlsChannelCredentials) {
   GPR_ASSERT(channel_credentials != nullptr);
 }
 
+// This test demonstrates how the SPIFFE credentials will be used to create
```
s/SPIFFE/TLS/g

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/client/credentials_test.cc:
```diff
@@ -660,14 +669,30 @@ TEST_F(CredentialsTest, LoadTlsChannelCredentials) {
 
+// This test demonstrates how the SPIFFE credentials will be used to create
+// server credentials.
+TEST_F(CredentialsTest, LoadSpiffeServerCredentials) {
```
s/SPIFFE/TLS/

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/end2end/end2end_test.cc:
```diff
@@ -256,7 +257,9 @@ void TestScenario::Log() const {
 class End2endTest : public ::testing::TestWithParam<TestScenario> {
  protected:
   static void SetUpTestCase() { grpc_init(); }
+
```
Please remove the empty lines.

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/end2end/end2end_test.cc:
```diff
@@ -2175,6 +2182,7 @@ std::vector<TestScenario> CreateTestScenarios(bool use_proxy,
 #endif
 
   if (test_secure) {
+    EnableSpiffeCredentials(GetCredentialsProvider());
```
s/Spiffe/Tls/g

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/end2end/end2end_test.cc:
```diff
@@ -405,6 +410,7 @@ class End2endTest : public ::testing::TestWithParam<TestScenario> {
 
     stub_ = grpc::testing::EchoTestService::NewStub(channel_);
     DummyInterceptor::Reset();
+    WaitOnServerAuthorizationToComplete(GetCredentialsProvider());
```
It seems `ResetStub` API will be triggered for non-TLS cases as well in which case we do not need server authorization.  Is there a way to avoid `WaitOnServerAuthorizationToComplete` call on the path common to all credentials?

matthewstevenson88 replied:
I added a `credentials_type` parameter to `WaitOnSpawnedThreads`. If `credentials_type` is anything other than `kSpiffeCredentialsType`, then it immediately returns.

yihuazhang commented on test/cpp/end2end/end2end_test.cc:
```diff
@@ -2175,6 +2182,7 @@ std::vector<TestScenario> CreateTestScenarios(bool use_proxy,
 #endif
 
   if (test_secure) {
+    EnableSpiffeCredentials(GetCredentialsProvider());
```
It seems all other credential types are obtained via `GetCredentialsProvider()->GetSecureCredentialsTypeList()`. Can we do the same for TLS?

matthewstevenson88 replied:
Done. This caused other e2e tests (e.g. `PortSharingEnd2endTest`) to start using the SPIFFE credentials, which necessitated some changes to `tls_test_credentials.{h,cc}`. Namely, I needed to introduce the `TlsThread` and `TlsData` objects. PTAL when I post the next commit.

yihuazhang commented on test/cpp/util/spiffe_test_credentials.cc:
```diff
@@ -0,0 +1,144 @@
+namespace grpc {
+namespace testing {
+
+class TestSyncTlsCredentialReload
```
Please rename the function name to `tls_test_credentials.cc`.

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/util/spiffe_test_credentials.cc:
```diff
@@ -0,0 +1,144 @@
+      test_credential_reload_config(
+          new ::grpc_impl::experimental::TlsCredentialReloadConfig(
+              credential_reload_interface));
+
```
Please remove empty lines within the function.

matthewstevenson88 replied:
Done in next commit.

yihuazhang commented on test/cpp/util/spiffe_test_credentials.h:
```diff
@@ -0,0 +1,56 @@
+namespace grpc {
+namespace testing {
+
+const char kSpiffeCredentialsType[] = "tls";
```
Please also rename the file to `tls_test_credentials.h`.

matthewstevenson88 replied:
Done. But note that I can't rename `kSpiffeCredentialsType` to `kTlsCredentialsType`, because you already renamed the latter from `kSslCredentialsType`.

yihuazhang commented on test/cpp/util/test_credentials_provider.h:
```diff
@@ -68,6 +68,11 @@ class CredentialsProvider {
   // Provide a list of secure credentials type.
   virtual std::vector<grpc::string> GetSecureCredentialsTypeList() = 0;
+
+  bool IsDefault() { return is_default_; }
```
What does `default` mean here?

matthewstevenson88 replied:
This refers to whether we are using the `DefaultCredentialsProvider`. (Unless one provides an alternate implementation of this interface, the `GetCredentialsProvider` method will return a `DefaultCredentialsProvider` instance.) I've changed the variable and method names to make this more clear.

matthewstevenson88 commented:
Thanks for the review @yihuazhang! I have some further comments below:

1. I resolved the comments you left from a review in late November, since they were about a now outdated version.
2. All references to "spiffe" should be removed now, except for the `kSpiffeCredentialsType` constant defined in `test_credentials_provider.h`. The reason that this constant cannot be changed is because there is already a `kTlsCredentialsType` declared in the same header file (it was formerly `kSslCredentialsType`, but it got changed recently).
3. Removing the `EnableSpiffeCredentials` API caused the TLS credentials to be used in additional end-to-end tests. In order for these tests to pass, I needed to introduce the `TlsThread` and `TlsData` classes, defined in `tls_test_credentials.h`. One example use-case that they handle is when the same credentials are used to create multiple channels (in which case, several server authorization checks may occur).
4. For the API's that I have added to the `CredentialsProvider` interface and the `DefaultCredentialsProvider` class, I have renamed them in a more generic way. When you approve this PR, I'll add Mark as a reviewer so he can approve those changes.

yihuazhang commented on include/grpcpp/security/tls_credentials_options.h:
```diff
@@ -56,7 +56,7 @@ class TlsKeyMaterialsConfig {
 
   /** Setter for key materials that will be called by the user. The setter
-   * transfers ownership of the arguments to the config. **/
+   *  takes ownership of the arguments to the config. **/
```
Please revise this comment saying that the ownership will be transferred.

matthewstevenson88 replied:
Per our offline discussion, this comment is no longer needed.

yihuazhang commented on src/cpp/common/tls_credentials_options.cc:
```diff
@@ -33,7 +31,7 @@ void TlsKeyMaterialsConfig::set_pem_root_certs(grpc::string pem_root_certs) {
 
 void TlsKeyMaterialsConfig::add_pem_key_cert_pair(
```
It seems that the ownership of key materials that get passed in to `TlsKeyMaterialsConfig` and `TlsCredentialReloadArg` are treated differently. In the former, the ownership is transferred while in the latter the ownership is not transferred. Since in both cases the key materials are input from users, should not we have a consistent behavior?

matthewstevenson88 replied:
For consistency, ownership will not be transferred in either case.

yihuazhang commented on test/cpp/client/credentials_test.cc:
```diff
@@ -670,6 +679,22 @@ TEST_F(CredentialsTest, LoadTlsChannelCredentials) {
   GPR_ASSERT(channel_credentials != nullptr);
 }
 
+// This test demonstrates how the TLS credentials will be used to create
```
Can we have a channel-counterpart for this test as well?

yihuazhang replied:
You probably need to re-sync to upstream in order to do that due to recent changes by Microsoft.

yihuazhang commented on test/cpp/util/tls_test_credentials.h:
```diff
@@ -0,0 +1,91 @@
+    }
+  }
+
+  std::thread thread_;
```
Please use `grpc_core::Thread` instead.

matthewstevenson88 replied:
Just to check: I should still use `grpc_core::Thread` even if this isn't in the C-core? I thought any C++11 API's were ok, at least in the C++ wrapper.

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/util/tls_test_credentials.cc:
```diff
@@ -0,0 +1,140 @@
+namespace grpc {
+namespace testing {
+
+class TestSyncTlsCredentialReload
```
s/TestSyncTlsCredentialReload/TestTlsSyncCredentialReload

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/util/tls_test_credentials.cc:
```diff
@@ -0,0 +1,140 @@
+  }
+};
+
+class TestSyncTlsServerAuthorizationCheck
```
s/TestSyncTlsServerAuthorizationCheck/TestTlsSyncServerAuthorizationCheck/

matthewstevenson88 replied:
Done.

yihuazhang commented on test/cpp/util/tls_test_credentials.cc:
```diff
@@ -0,0 +1,140 @@
+  arg->OnServerAuthorizationCheckDoneCallback();
+}
+
+class TestAsyncTlsServerAuthorizationCheck
```
s/TestAsyncTlsServerAuthorizationCheck/TestTlsAsyncServerAuthorizationCheck

matthewstevenson88 replied:
Done.

matthewstevenson88 commented:
Thanks for the review @yihuazhang! I've addressed the changes in the newest commit.

yihuazhang commented on include/grpcpp/security/tls_credentials_options.h:
```diff
@@ -55,12 +55,12 @@ class TlsKeyMaterialsConfig {
-  /** Setter for key materials that will be called by the user. The setter
-   * transfers ownership of the arguments to the config. **/
-  void set_pem_root_certs(grpc::string pem_root_certs);
+  /** Setter for key materials that will be called by the user. **/
```
Please add some comments about the ownership of key materials.

matthewstevenson88 replied:
Done.

yihuazhang commented on include/grpcpp/security/tls_credentials_options.h:
```diff
@@ -85,25 +85,20 @@ class TlsCredentialReloadArg {
-   * original key_materials_config or error_details after using the appropriate
-   * setter function.
-   * **/
+  /** Setters for member fields. **/
```
Please also add some comments about the ownership of key materials.

matthewstevenson88 replied:
Done.

yihuazhang commented on src/cpp/common/tls_credentials_options.cc:
```diff
@@ -86,16 +84,43 @@ void TlsCredentialReloadArg::set_pem_root_certs(
+    const TlsKeyMaterialsConfig::PemKeyCertPair& pem_key_cert_pair) {
   grpc_ssl_pem_key_cert_pair* ssl_pair =
       (grpc_ssl_pem_key_cert_pair*)gpr_malloc(
           sizeof(grpc_ssl_pem_key_cert_pair));
```
Sorry for bringing this up late. How about refactor it (i.e., create a helper function in an anonymous namespace) so that it can be used in  `TlsCredentialReloadArg::set_key_materials` as well?

matthewstevenson88 replied:
Done.

yihuazhang commented:
Looks pretty good. Only some minor comments are left.

yihuazhang commented on include/grpcpp/security/tls_credentials_options.h:
```diff
@@ -55,7 +55,8 @@ class TlsKeyMaterialsConfig {
 
-  /** Setter for key materials that will be called by the user. **/
+  /** Setter for key materials that will be called by the user. They do not take
+   *  ownership of the arguments. **/
```
Maybe rephrase it to "ownership of the arguments will not be transferred". It's unclear if what `they` refers to.

matthewstevenson88 replied:
Done.

yihuazhang commented on include/grpcpp/security/tls_credentials_options.h:
```diff
@@ -91,7 +92,8 @@ class TlsCredentialReloadArg {
   grpc::string error_details() const;
 
-  /** Setters for member fields. **/
+  /** Setters for member fields. They do not take ownership of the arguments.
```
Same as above.

matthewstevenson88 replied:
Done.

yihuazhang commented on src/cpp/common/tls_credentials_options.cc:
```diff
@@ -83,16 +83,24 @@ void TlsCredentialReloadArg::set_pem_root_certs(
-void TlsCredentialReloadArg::add_pem_key_cert_pair(
+namespace {
+
+::grpc_core::PemKeyCertPair convert_to_core_pem_key_cert_pair(
```
s/convert_to_core_pem_key_cert_pair/ConvertToCorePemKeyCertPair/g

matthewstevenson88 commented:
@markdroth Could you please have a look at the modifications to the end-to-end tests and the test credential provider? More specifically, the modifications to the following common test files:

- `test_credentials_provider.{h,cc}`
- `end2end_test.cc`, `async_end2end_test.cc`, `client_callback_end2end_test.cc`, `port_sharing_end2end_test.cc`, and `shutdown_test.cc`.

Also relevant are the test TLS credentials, which are generated in `tls_test_credentials.{h,cc}`.

A brief summary of the changes: the test TLS credentials perform an async server authorization check, and the changes to the test files are all related to managing the threads spawned by this server authorization check. @yihuazhang has approved the TLS-specific changes of this PR, and we would appreciate your OK for the changes to the above common test files.

Thanks! :)
